## Introduction

In this assignment we are using feed forward neural network to print an hexadecimal output from a given binary input.We are using neural network to replace a complex electric circuit to convert 4 inputs into an seven segment display which would save lot of time and energy.In this we will take 4 inputs and convert them into seven outputs to print it on a seven segment display.

How does a neural network work?  
This is a network of layers and nodes where each layer does a differnt task and the data which is fed into it is processed at each layer and that output is fed into the next one, and by the time it reaches the last layer it is been through many nodes and is been processed differently and has been altered at each node and layer,our brain also works in a similar way to derive a conclusion, when certain information is fed into it, we think from differnt perspectives and situations to come to a conclusion, as these both work in a similar fashion this technology got its name.

What is a Node?  
It is the basic unit of the neural network where a function is been applied, it takes multiple inputs with weights and applies the desired function based on the weights we assign to the each node and computes an output.

What is feed forward neural network?  
As the name suggests the input is carried from one node to the other, in which the input is processed at a node and the output of that node is carried forward and becomes the input for the next node and so on. In this network we use multiple layers, and the data gets processed at each layer.


In [142]:
#importing the necessary libraries
import matplotlib.pyplot as plt
import numpy, random, os

# keras libraries
from numpy import loadtxt
from keras.models import Sequential
from keras.layers import Dense,Dropout

In [143]:

# load the dataset
dataset = loadtxt('hexa_display.csv', delimiter=',')
# split into input (X) and output (y) variables here we are taking the first 4 columns as inputs and the rest as outputs
X = dataset[:,0:4]
y = dataset[:,4:]


In [144]:
#checking our data in y
y

array([[1., 1., 1., 1., 1., 1., 0.],
       [0., 1., 1., 0., 0., 0., 0.],
       [1., 1., 0., 1., 1., 0., 1.],
       [1., 1., 1., 1., 0., 0., 1.],
       [0., 1., 1., 0., 0., 1., 1.],
       [1., 0., 1., 1., 0., 1., 1.],
       [1., 0., 1., 1., 1., 1., 1.],
       [1., 1., 1., 0., 0., 0., 0.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 0., 1., 1.],
       [1., 1., 1., 0., 1., 1., 1.],
       [0., 0., 1., 1., 1., 1., 1.],
       [1., 0., 0., 1., 1., 1., 0.],
       [0., 1., 1., 1., 1., 0., 1.],
       [1., 0., 0., 1., 1., 1., 1.],
       [1., 0., 0., 0., 1., 1., 1.]])

Here we are creating our model with three layers and we are defining the number of nodes, type of function we are using in each layer and also using droupout which would help in preventing overfitting it restricts the use of some of the nodes randomly so it doesnt overfit the output.Here we are using activation layers as relu and sigmoid.The relu funtion will simply give the input if it is positive else it will give zero as output it also works faster than sigmoid function. Sigmoid function works in a similar fashion to logistic regression and gives the output as one or zero,we are using these functions in these layers because we want our outputs to be 1 or 0, for switching on our desired circuits.

In [145]:

# defining the keras model
model = Sequential()
#here we are setting the number of nodes, input parameters and the algorithm used for 1st layer
model.add(Dense(30, input_dim=4, activation='relu'))
#it is used to prevent overfitting
model.add(Dropout(0.2))
#we are using 22 hidden nodes with relu function, 2nd layer
model.add(Dense(22, activation='relu'))
model.add(Dropout(0.2))
#we are using 7 nodes as we only need 7 outputs and sigmoid as activation function as we need our output to be 1 or 0
model.add(Dense(7, activation='sigmoid'))

In [146]:
# compile the keras model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

Batch size and epoch are hyperparameters in which both work as for loop to itetrate through the data and batch size acts as a nested for loop inside the epoch for loop, here in our case we are running 4 times the same data through all the nodes to get the output and each time we are running it we are running through the whole batch size. I.e we are running the whole data in each batch before updating it.

In [165]:
# fit the keras model on the dataset
model.fit(X, y, epochs=4, batch_size=16,validation_split=0.2)

Train on 12 samples, validate on 4 samples
Epoch 1/4
12/12 [==============================] - 0s 1ms/step - loss: 0.3283 - accuracy: 0.8690 - val_loss: 1.6263 - val_accuracy: 0.6071
Epoch 2/4
12/12 [==============================] - 0s 463us/step - loss: 0.2475 - accuracy: 0.9048 - val_loss: 1.6291 - val_accuracy: 0.6071
Epoch 3/4
12/12 [==============================] - 0s 741us/step - loss: 0.2630 - accuracy: 0.9048 - val_loss: 1.6318 - val_accuracy: 0.6071
Epoch 4/4
12/12 [==============================] - 0s 744us/step - loss: 0.2898 - accuracy: 0.8929 - val_loss: 1.6347 - val_accuracy: 0.6071


We should be careful while selecting the number of epochs, as after a certain number of epochs our accuracy starts dropping and we have to figure out where it is changing and select the best epochs to get more accurcy.

In [166]:
# evaluating our model
_, accuracy = model.evaluate(X, y)
print('Accuracy: %.2f' % (accuracy*100))

16/16 [==============================] - 0s 198us/step
Accuracy: 83.04


In [171]:
predictions = model.predict(X)

In [172]:
# here we are checking our output values
predictions

array([[0.8351643 , 0.81571984, 0.8940244 , 0.595598  , 0.69683194,
        0.871526  , 0.6116091 ],
       [0.50875044, 0.80298156, 0.97878945, 0.3362398 , 0.03316364,
        0.21599287, 0.19724822],
       [0.9471343 , 0.85934514, 0.3733492 , 0.785686  , 0.88580966,
        0.28286064, 0.8549864 ],
       [0.8723745 , 0.82034075, 0.91614854, 0.7461647 , 0.12427038,
        0.04724562, 0.6804404 ],
       [0.34462982, 0.73749197, 0.9990897 , 0.34333783, 0.11610115,
        0.99243224, 0.91395795],
       [0.63181806, 0.5304072 , 0.99747384, 0.6013276 , 0.03175643,
        0.86566097, 0.78991604],
       [0.7991234 , 0.5016501 , 0.9401224 , 0.6463005 , 0.82940304,
        0.91526747, 0.915884  ],
       [0.81609505, 0.65768903, 0.963698  , 0.5857509 , 0.04611129,
        0.11092857, 0.47747996],
       [0.9753566 , 0.9224229 , 0.99515104, 0.86519265, 0.985405  ,
        0.9996425 , 0.9946284 ],
       [0.69643575, 0.8567632 , 0.9992404 , 0.8088832 , 0.23694941,
        0.9866371 , 0.9

Here we are converting our outputs into 1 or 0 based on 0.5 threshold as we need the output to be 1 or 0 to turn on a circuit equally bright.

In [173]:
#converting our predictions
B = numpy.where(predictions > 0.5, 1, 0)

In [174]:
#predicted data
B

array([[1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 0, 0, 0, 0],
       [1, 1, 0, 1, 1, 0, 1],
       [1, 1, 1, 1, 0, 0, 1],
       [0, 1, 1, 0, 0, 1, 1],
       [1, 1, 1, 1, 0, 1, 1],
       [1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 0, 0, 0],
       [1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 0, 1, 1],
       [1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 0, 1, 1],
       [1, 1, 1, 1, 1, 1, 1],
       [1, 0, 1, 1, 0, 1, 1]])

In [175]:
#actual data
y

array([[1., 1., 1., 1., 1., 1., 0.],
       [0., 1., 1., 0., 0., 0., 0.],
       [1., 1., 0., 1., 1., 0., 1.],
       [1., 1., 1., 1., 0., 0., 1.],
       [0., 1., 1., 0., 0., 1., 1.],
       [1., 0., 1., 1., 0., 1., 1.],
       [1., 0., 1., 1., 1., 1., 1.],
       [1., 1., 1., 0., 0., 0., 0.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 0., 1., 1.],
       [1., 1., 1., 0., 1., 1., 1.],
       [0., 0., 1., 1., 1., 1., 1.],
       [1., 0., 0., 1., 1., 1., 0.],
       [0., 1., 1., 1., 1., 0., 1.],
       [1., 0., 0., 1., 1., 1., 1.],
       [1., 0., 0., 0., 1., 1., 1.]])

## Conclusion
Our model seems to work well and we got an accuracy of 83.04% which is good, and this whole model took very less time compared to building the whole circuit manually and also saved lot of time and we can further try to improve the model and we also do not get affected by circuitry mistakes or IC blowouts due to which our model is more trustworthy. We also learned to train and build an neural network on the way to solve our problem.We also got rid of spending money on buying all the components, the output we are generating is also very stable and bright.

## Refernces


Brownlee, J. (2019, August 6). A Gentle Introduction to the Rectified Linear Unit (ReLU). Retrieved from https://machinelearningmastery.com/rectified-linear-activation-function-for-deep-learning-neural-networks/.  

Brownlee, J. (2019, October 25). Difference Between a Batch and an Epoch in a Neural Network. Retrieved from https://machinelearningmastery.com/difference-between-a-batch-and-an-epoch/.